lnst/RecipeCommon/Perf/Measurements/LinuxPerfMeasurement.py

```diff
@@ -0,0 +1,240 @@
+
+    def start(self) -> None:
+        self._start_timestamp = time.time()
+        for host, filename, cpus in self._generate_configurations():
+            self._jobs.append(host.run(LinuxPerf(output_file=filename, cpus=cpus), bg=True))
+
+    def finish(self) -> None:
+        for job, (host, _, _) in zip(self._jobs, self._generate_configurations()):
+            job.kill(signal=signal.SIGINT)
+            if not job.wait(timeout=120):
+                logging.error(f"timeout while waiting for linuxperf job to finish on host {host.hostid}")
+        self._end_timestamp = time.time()
+
+    def collect_results(self) -> list[BaseMeasurementResults]:
+        self._collection_index += 1
+        results: list[BaseMeasurementResults] = []
+        for job, (host, filename, cpus) in zip(self._jobs, self._generate_configurations()):
+            if job.result is None:
+                continue
+
+            # copy agent's perf.data file to a controller
+            src_filepath: str = job.result["filename"]
+            new_filename: str = f"{filename}.{self._collection_index}.{timestamp()}"
+            dst_filepath: str = os.path.join(self._data_folder, host.hostid, new_filename)
+
+            host.copy_file_from_machine(src_filepath, dst_filepath)
+            logging.debug(f"perf-record data copied from agent to {dst_filepath}")
+            results.append(
+                LinuxPerfMeasurementResults(
+                    self,
+                    host,
+                    dst_filepath,
+                    cpus,
+                    self._start_timestamp,
+                    self._end_timestamp,
+                )
+            )
+        return results
+
+    @classmethod
+    def report_results(
+        cls,
+        recipe: BaseRecipe,
+        aggregated_results: list[AggregatedLinuxPerfMeasurementResults],
+    ):
+        for aggregated_result in aggregated_results:
+            cpus: str = ",".join(map(str, aggregated_result.individual_results[0].cpus))
+            hostid: str = aggregated_result.individual_results[0].host.hostid
+            files: str = "\n    ".join(
+                result.filename for result in aggregated_result.individual_results
+            )
+
+            recipe.add_result(
+                True,
+                f"perf-record recorded CPU(s) {cpus} on {hostid} to files:\n    {files}",
+            )
+
+    @classmethod
+    def aggregate_results(
+        cls,
+        old: Optional[list[AggregatedLinuxPerfMeasurementResults]],
+        new: list[LinuxPerfMeasurementResults],
+    ) -> list[AggregatedLinuxPerfMeasurementResults]:
+        if old is None:
+            return [AggregatedLinuxPerfMeasurementResults(result) for result in new]
+
+        aggregated: list[AggregatedLinuxPerfMeasurementResults] = []
+        for old_measurements, new_measurement in zip(old, new):
+            aggregated.append(
+                cls._aggregate_linux_perf_results(old_measurements, new_measurement)
```

Comment:
I think the code between lines 226-230 can be simplified:
```
for old_measurements, new_measurement in zip(old, new):
    old_measurements.add_results(new)
```

And that's it. You don't need `_aggregate_linux_perf_results()` at all, imo.

If I'm missing something please let me know. Here's my code flow analysis:
```
perf_test()
    perf_test_iteration()
        """
        for measurement in recipe_conf.measurements:
            measurement.start()
        for measurement in reversed(recipe_conf.measurements):
            measurement.finish()
        for measurement in recipe_conf.measurements:
            measurement_results = measurement.collect_results()
            results.add_measurement_results(
                measurement, measurement_results
        """

        measurement_results = measurement.collect_results()
            -> measurement_results is a list of LinuxPerfMeasurementResults
            [
                # iteration 1, all hosts, all cpus
                LinuxPerfMeasurementResults(host1, cpu0),
                LinuxPerfMeasurementResults(host2, cpu0),
                LinuxPerfMeasurementResults(host1, cpu5),
                ...
            ]

        results.add_measurement_results(measurement, measurement_results)
        ->
            appends individual measurement results for the measurement, then
            aggregates:
                measurement.aggregate_results()
                    first iteration:
                        old=None, new=measurement_results
                        [
                            AggregatedLinuxPerfMeasurementResults(LinuxPerfMeasurementResults(host1, cpu0)),
                            AggregatedLinuxPerfMeasurementResults(LinuxPerfMeasurementResults(host2, cpu0)),
                            AggregatedLinuxPerfMeasurementResults(LinuxPerfMeasurementResults(host1, cpu5)),
                            ...
                        ]
                    next iteration:
                        old=previous, new=measurement_results
                        aggregated=[
                            AggregatedLinuxPerfMeasurementResults:[
                                LinuxPerfMeasurementResults(host1, cpu0) # first iteration
                                LinuxPerfMeasurementResults(host1, cpu0) # second iteration
                            ],
                            AggregatedLinuxPerfMeasurementResults:[
                                LinuxPerfMeasurementResults(host2, cpu0) # first iteration
                                LinuxPerfMeasurementResults(host2, cpu0) # second iteration
                            ],
                            AggregatedLinuxPerfMeasurementResults:[
                                LinuxPerfMeasurementResults(host1, cpu5) # first iteration
                                LinuxPerfMeasurementResults(host1, cpu5) # second iteration
                            ], ...
                        ]
```

Reply:
You're right. This was done to mimic `BaseFlowMeasurement` and `BaseCPUMeasurement`, which both use two functions. I will simplify it, since there's no logic needed in the child function.

Reply:
resolved in 18a13800